network/src/protocols/outbound_peer.rs

```diff
@@ -51,21 +55,24 @@ impl OutboundPeerService {
             .take(count as usize)
         {
             debug!(target: "network", "dial attempt peer: {:?}", addr);
-            self.network_state
-                .dial_all(&mut p2p_control, &peer_id, addr);
+            self.network_state.dial_all(&p2p_control, &peer_id, addr);
         }
     }
 
     fn feeler_peers(&mut self, count: u32) {
-        let peers = self.network_state.peer_store().peers_to_feeler(count);
-        let mut p2p_control = self.p2p_control.clone();
+        let peers = self
+            .network_state
+            .with_peer_store(|peer_store| peer_store.peers_to_feeler(count));
+        let p2p_control = self.p2p_control.clone();
         for (peer_id, addr) in peers
             .into_iter()
             .filter(|(peer_id, _addr)| self.network_state.local_peer_id() != peer_id)
         {
+            self.network_state.with_peer_registry_mut(|reg| {
+                reg.add_feeler(peer_id.clone());
```

Comment:
What if dial failed?

PeerRegistry should only store peers which already connected, move this into 
 `Feeler.connected`.